numba/core/event.py

```diff
@@ -0,0 +1,357 @@
+
+    Parameters
+    ----------
+    kind : str
+    listener : Listener
+    """
+    assert isinstance(listener, Listener)
+    kind = _guard_kind(kind)
+    lst = _registered[kind]
+    lst.remove(listener)
+
+
+def broadcast(event):
+    """Broadcast an event to all registered listeners.
+
+    Parameters
+    ----------
+    event : Event
+    """
+    for listener in _registered[event.kind]:
+        listener.notify(event)
+
+
+class Listener(abc.ABC):
+    """Base class for all event listeners.
+    """
+    @abc.abstractmethod
+    def on_start(self, event):
+        """Called when there is a *START* event.
+
+        Parameters
+        ----------
+        event : Event
+        """
+        pass
+
+    @abc.abstractmethod
+    def on_end(self, event):
+        """Called when there is a *END* event.
+
+        Parameters
+        ----------
+        event : Event
+        """
+        pass
+
+    def notify(self, event):
+        """Notify this Listener with the given Event.
+        """
+        if event.is_start:
+            self.on_start(event)
+        elif event.is_end:
+            self.on_end(event)
+        else:
+            raise AssertionError("unreachable")
+
+
+class TimingListener(Listener):
+    """A listener that measures the duration between a *START* event and
+    its matching *END* event.
+    """
+    def __init__(self):
+        self._depth = 0
+
+    def on_start(self, event):
+        if self._depth == 0:
+            self._ts = timer()
+        self._depth += 1
+
+    def on_end(self, event):
+        self._depth -= 1
+        if self._depth == 0:
+            self._duration = timer() - self._ts
+
+    @property
+    def done(self):
+        """Returns a ``bool`` indicating a measurement is made.
+
+        When this returns ``False``, the matching event has never fired.
+        Iff this returns ``True``, ``.duration`` can be read without error.
+        """
+        return hasattr(self, "_duration")
+
+    @property
+    def duration(self):
+        """Returns the measured duration.
+
+        This may raise ``AttributeError``. Users can use ``.done`` to check
+        that a measurement has been made.
+        """
+        return self._duration
+
+
+class RecordingListener(Listener):
+    """A listener that records all event and store it in the ``.buffer``
+    attribute as a list of 2-tuple ``(float, Event)``, where the first element
+    is the time of the event as returned by ``time.time()`` and the second
+    element is the event.
+    """
+    def __init__(self):
+        self.buffer = []
+
+    def on_start(self, event):
+        self.buffer.append((time.time(), event))
+
+    def on_end(self, event):
+        self.buffer.append((time.time(), event))
+
+
+@contextmanager
+def install_listener(kind, listener):
+    """Install a listener temporarily within the duration of the context.
+
+    Returns
+    -------
+    res : Listener
+        The *listener* provided.
```

Comment:
It may be worthwhile adding a simple example of how to use this?

Reply:
Maybe also for the other context managers below?

Reply:
Docs added in https://github.com/numba/numba/pull/6626/commits/1b814b86df9c9b272c32643999a88ba75b66245b

Reply:
Great!